### PR TITLE
Upgrade csharp Pester and PesterTest projects, add tests

### LINKED ISSUE #1861: Upgrade csharp projects and dependencies, utilize language options
## Summary of the feature request 

Pester should use later versions of .net framework and upgrade csharp dependencies.  Pester should check for null values in code at compile-time.

## How should it work? (optional) 

```xml
    <LangVersion>8.0</LangVersion>
    <Nullable>enable</Nullable>
    <TargetFrameworks>netstandard2.1;net48</TargetFrameworks>
    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```

See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types for nullable type checks.  Nullable check is safe to implement.

Upgrade tests could be done as part of devops test matrix and implemented on a branch.

```shell
# Pester
dotnet add package System.Management.Automation

# Pester.Tests
dotnet add package xunit
dotnet add package Microsoft.NET.Test.Sdk
dotnet add package xunit.runner.visualstudio
dotnet add package coverlet.collector
```

Wondering if changing target framework would break any downstream consumers or devs?
Is there any need for additional test coverage of the C# code?

### PR DESCRIPTION
**Pull request by asears**
## PR Summary

_This awaits integration of code coverage PR_

Upgrade target framework for csharp projects.
Upgrade dependencies.
Add vscode and editorconfig support.
Add tests.

Fix #1861

## PR Checklist

- [X] PR has meaningful title
- [X] Summary describes changes
- [ ] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [X] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

### CODE REVIEW

File: src/csharp/PesterTests/ContainerInfoTests.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+namespace PesterTests
+{
+    using Pester;
```
Comment (nohwnd) [inline src/csharp/PesterTests/ContainerInfoTests.cs:3]: Usings in namespace, is a style choice I don't like.

File: src/csharp/PesterTests/ContainerInfoTests.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    using Xunit;
+
+    /// <summary>
+    /// Container Info Tests
```
Comment (nohwnd) [inline src/csharp/PesterTests/ContainerInfoTests.cs:7]: Sorry, but those annotations are just noise.

File: src/csharp/PesterTests/FactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        [Fact]
+        /// <summary>
+        /// Test CreateNoteProperty
+        /// </summary>
+        public void Test_CreateNoteProperty()
+        {
+            var name = "test";
+            var value = "test";
+
+            var actual = Factory.CreateNoteProperty(name, value);
```
Comment (nohwnd) [inline src/csharp/PesterTests/FactoryTests.cs:21]: Nothing is asserted in those tests.

File: src/csharp/Pester/Pester.csproj [hunk 4]
```diff
@@ -1,14 +1,17 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net452</TargetFrameworks>
+    <LangVersion>8.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <TargetFrameworks>netstandard2.1;net48</TargetFrameworks>
```
Comment (nohwnd) [inline src/csharp/Pester/Pester.csproj:6]: Using net452 is deliberate to be compatible with powershell 3.

File: src/csharp/PesterTests/InvocationResultTests.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+        public InvocationResultTests()
+        {
+            var success = true;
+            var errorRecords = new List<ErrorRecord>();
+            var standardOutput = new Object();
+
+            InvocationResult testInvocationResult =
+                new InvocationResult(success, errorRecords, standardOutput);
+            Assert.NotNull (testInvocationResult);
```
Comment (nohwnd) [inline src/csharp/PesterTests/InvocationResultTests.cs:24]: Please no spaces after method name.

File: src/csharp/PesterTests/PesterTests.csproj [hunk 6]
```diff
@@ -1,16 +1,25 @@
...
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0" />
-    <PackageReference Include="xunit" Version="2.4.0" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
-    <PackageReference Include="coverlet.collector" Version="1.2.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.1" />
```
Comment (nohwnd) [inline src/csharp/PesterTests/PesterTests.csproj:12]: I'd stick with the older test.sdk, the new one is buggy. 16.7.1 if you must upgrade.

File: src/csharp/PesterTests/TestTests.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    using Pester;
+    using Xunit;
+
+    /// <summary>
+    /// Test the test class
+    /// </summary>
+    public class TestTests
+    {
+        public readonly Test test;
+        public TestTests() => test = new Test();
```
Comment (nohwnd) [inline src/csharp/PesterTests/TestTests.cs:12]: Why not just initialize it in the field initializer? Or write it in the test?

File: src/csharp/PesterTests/ContainerInfoTests.cs (hunk 1 above)
Comment (asears) [reply src/csharp/PesterTests/ContainerInfoTests.cs:3]: I got some warning around this, prefer the other convention too.  I'll revert as it doesn't matter in terms of functionality.  Using Roslynator to scan.

File: src/csharp/PesterTests/ContainerInfoTests.cs (hunk 2 above)
Comment (asears) [reply src/csharp/PesterTests/ContainerInfoTests.cs:7]: I'll silence the linter in other ways and remove the xml comments.

File: src/csharp/Pester/Pester.csproj (hunk 4 above)
Comment (asears) [reply src/csharp/Pester/Pester.csproj:6]: Thanks for clarification.  Will revert

File: src/csharp/PesterTests/InvocationResultTests.cs (hunk 5 above)
Comment (asears) [reply src/csharp/PesterTests/InvocationResultTests.cs:24]: Will see where the autoformatter is falling down here.

File: src/csharp/PesterTests/FactoryTests.cs (hunk 3 above)
Comment (asears) [reply src/csharp/PesterTests/FactoryTests.cs:15]: Implemented a test, further tests required here.

File: src/csharp/PesterTests/TestTests.cs (hunk 7 above)
Comment (asears) [reply src/csharp/PesterTests/TestTests.cs]: wrote it in the test.

File: src/csharp/PesterTests/PesterTests.csproj (hunk 6 above)
Comment (asears) [reply src/csharp/PesterTests/PesterTests.csproj]: reverted

File: src/csharp/PesterTests/InvocationResultTests.cs (hunk 5 above)
Comment (asears) [reply src/csharp/PesterTests/InvocationResultTests.cs]: Had prettier enabled from another PR.

**Comment by asears** [comment]
@nohwnd which branch should I be targeting here? I may need to clean up the merge / revert some files here.  Coverage test still failing.  

```
PesterTests.CoverageTests.Test_ToString_ShouldReturnPercentStringForValidDecimal

Expected: 100%
Actual:   N2 %
```

**Comment by nohwnd** [comment]
@asears  merge v5.0

**Pull request closed without merging**